go-controller/pkg/ovn/loadbalancer.go

```diff
@@ -82,45 +84,170 @@ func (ovn *Controller) getLoadBalancerVIPS(
 	return raw, nil
 }
 
+// deleteLoadBalancerVIP removes the VIP as well as any reject ACLs associated to the LB
 func (ovn *Controller) deleteLoadBalancerVIP(loadBalancer, vip string) {
 	vipQuotes := fmt.Sprintf("\"%s\"", vip)
-	stdout, stderr, err := util.RunOVNNbctl("--if-exists", "remove",
-		"load_balancer", loadBalancer, "vips", vipQuotes)
+	stdout, stderr, err := util.RunOVNNbctl("--if-exists", "remove", "load_balancer", loadBalancer, "vips", vipQuotes)
 	if err != nil {
 		klog.Errorf("Error in deleting load balancer vip %s for %s"+
 			"stdout: %q, stderr: %q, error: %v",
 			vip, loadBalancer, stdout, stderr, err)
+		// if we hit an error and fail to remove load balancer, we skip removing the rejectACL
+		return
 	}
+	ovn.removeServiceEndpoints(loadBalancer, vip)
+	ovn.deleteLoadBalancerRejectACL(loadBalancer, vip)
+	ovn.removeServiceLB(loadBalancer, vip)
 }
 
-func (ovn *Controller) createLoadBalancerVIP(lb string, serviceIP string, port int32, ips []string, targetPort int32) error {
+func (ovn *Controller) configureLoadBalancer(lb, serviceIP string, port int32, endpoints []string) error {
+	commaSeparatedEps := strings.Join(endpoints, ",")
+	target := fmt.Sprintf(`vips:"%s"="%s"`, util.JoinHostPortInt32(serviceIP, port), commaSeparatedEps)
+
+	out, stderr, err := util.RunOVNNbctl("set", "load_balancer", lb,
+		target)
+	if err != nil {
+		klog.Errorf("Error in configuring load balancer: %s "+
+			"stdout: %q, stderr: %q, error: %v", lb, out, stderr, err)
+		ovn.setServiceEndpointsToLB(lb, util.JoinHostPortInt32(serviceIP, port), endpoints)
```

Comment:
this should happen when there is no error NOT when there is an error

Reply:
@dcbw Fixing that and moving locking for ACL/LB create up a bit fixes the problem. The race happens because ovn-nbctl commands take a long time and we were only locking when we update the map structure. This creates a race where endpoint thread is waiting for ovn-nbctl to update before locking the map, while the service has already checked the map, and started creating ACL. Fix is to move locking before the relevant nbctl commands are issued (basically lock around the entire create/configure operations).

Reply:
@trozet shouldn't the endpoint thread and the service thread basically generate the same ACL though?  eg they will actually get serialized already in nbctl, but even if one thinks the ACL isn't created yet becuase the other is creating it at that moment, wouldn't the second one just try to create the same ACL again a few seconds later and succeed?

Reply:
They do generate the same ACL, but in this the issue is when there should not be an ACL and a load balancer should exist. The issue is when a service gets created with a selector (and that selector corresponds to valid endpoints) the service and endpoint creation events will happen at almost exactly the same time. When that happens the service watcher fires and checks to see if an entry is in the map, and if not thinks there are no endpoints so it creates the ACL. On the other hand the endpoint watcher fires and tries to update the map with endpoints, but the issue is the service may have already checked and created the ACL before the endpoints watcher had done its part. In the endpoint watcher we were removing any reject ACL before we created the load balancer. I changed that to be after - to ensure that it does get removed by the endpoint watcher if service creates it before, and also improved the locking.